stan/math/rev/core/grad.hpp

```diff
@@ -27,13 +27,14 @@ namespace math {
  * @param vi Variable implementation for root of partial
  * derivative propagation.
  */
-static void grad(vari* vi) {
+template <typename Vari>
+static void grad(Vari* vi) {
   vi->init_dependent();
-  std::vector<vari*>& var_stack = ChainableStack::instance_->var_stack_;
-  size_t end = var_stack.size();
+  size_t end = ChainableStack::instance_->var_stack_.size();
   size_t beginning = empty_nested() ? 0 : end - nested_size();
   for (size_t i = end; i-- > beginning;) {
-    var_stack[i]->chain();
+    boost::variant2::visit([&](auto& x) { x->chain(); },
+                           ChainableStack::instance_->var_stack_[i]);
```

Comment:
@SteveBronder Do you have any explanation why this turs out to be faster than virtual calls?

Reply:
I ended up reverting this to the virtual call style. It's actually the same speed unless you use a custom visitor (since we know the most likely type is a `vari_value<double>*`) but idt this approach would have scaled well